include/io.h

```diff
@@ -1,52 +1,56 @@
-
 #pragma once
-
 #include <fstream>
 #include <string>
+#include <string_view>
 
-class IO
-{
+class IO {
     public:
-        explicit IO(const std::string& filename);
-        ~IO();
+        explicit IO(std::string_view filename); // Can throw
+        ~IO() noexcept;
 
         // Basic operations
-        char current() const { return current_char_; }
-        void next();
-        bool eof() const { return current_char_ == '\0' || file_stream_.eof(); }
+        [[nodiscard]] constexpr char current() const noexcept {
+            return current_char_;
+        }
+        void next() noexcept;
+        [[nodiscard]] constexpr bool eof() const noexcept {
+            return current_char_ == '\0' || file_stream_.eof();
```

Comment:
hmm, I just realized I dont think this is going to work anymore. `\0` is only at the end of C-strings, not `std::string` or `std::string_view`, and has a really bad C smell to it that does not fit with C++.

This is where a dynamic `std::string` makes a lot of sense. Then `current_char` will be the iterator or position in the `std:string` to the current character

I think for your learning you should use `std::iterator` for current position, each string has an iterator that is like a pointer that you can save.

Reply:
See also https://en.cppreference.com/w/cpp/string/basic_string/begin
and https://en.cppreference.com/w/cpp/string/basic_string/end

`current_char` and your position in the string will begin with `.begin()` and then you 
can use pointer arithmetic, and you check if it's at the end with `current_pos == string.end()`

Example:

```C++
 std::string str{"Hello world"};
  for (auto it = str.begin(); it != str.end(); it++){
  // notice the it++, "it" means iterator
  }
 ```